### PR TITLE
Implement nbShow and highlight.js support for code in markdown

### LINKED ISSUE #68: support for highlighting other languages
with incoming PR #64 we will start to see more languages in nimib.

we should make again easily available the use of highlight.js in documents which contain other languages (initially nimib did nim language highlighting through highlight.js, then thanks to #15, it passed to do it statically).

this would also ease the support of nim highlighting in pure markdown pages (such as those used in nimibook, see examples there) which is not available statically (it would require patching markdown generator or postprocess the html)

implementation should be rather straightforward. among (minor) steps to be understood how to do:
- [ ] how do we avoid highlight.js to work again on nim code blocks already highlighted?
- [ ] which highlight.js set of languages we will provide? (nim is not in any of the official packages, we could go with nim + common languages but this does not include Julia, which we would want to have) see here: https://highlightjs.org/download/)

### LINKED ISSUE #74: nbShow for images and tables
we should have a dedicated command nbShow whose behaviour will depend on the type of object is applied.

For example:
- when applied to a ggplot, it should show the image (would it be possible/advisable to save as svg and incorporate directly in html?)
- when applied to a dataframe, it should print a nice table

this should be in some nimib/show module that should uses `when compiles(import ggplotnim)` or similar mechanism not to introduce additional dependencies.

I plan to do this after #24 (hopefully I will start soon working on that, likely starting from scratch)

### PR DESCRIPTION
**Pull request by HugoGranstrom**
fixes #74 fixes #68 

- `nbShow` seems to work fine, see penguins.
- highlight.js seems to affect the statically highlighted code... So we have to solve that somehow.
  - E.g in penguins the `echo df` isn't highlighted correctly.

### CODE REVIEW

**Comment by HugoGranstrom** [comment]
One potential solution to the highlight.js problem is to give the user two modes:

1. Statically highlight Nim code but do not allow highlight.js to highlight Nim code (Nim code in markdown isn't highlighted).
2. Everything is highlighted using highlight.js.

**Comment by HugoGranstrom** [comment]
Nevermind, found that you can mark a code block as `nohighlight` :tada:

**Comment by pietroppeter** [comment]
Will look into it and let you know. As a first quick feedback I would probably keep `nim` in code class (not sure why I put hljs, was it needed for the css highlighting?)

**Comment by pietroppeter** [comment]
Some more feedback:
- should we make this something that we can opt out? It would require creating a new line in head partial and a new data point highlightJs in context that can be turned off. Putting next to the css feels a bit of a hack ;)
- do we have somewhere in the documents an example where we have Nim code in a markdown context, so that we can see it in action?

**Comment by pietroppeter** [comment]
(Btw not even sure if this is ready for review...)

**Comment by HugoGranstrom** [comment]
> As a first quick feedback I would probably keep nim in code class (not sure why I put hljs, was it needed for the css highlighting?)

In my minimal example adding the `nim` class doesn't make a difference when using `nohighlighting`, but adding `hljs` ruins it. So I could add `nim` back. 

> do we have somewhere in the documents an example where we have Nim code in a markdown context, so that we can see it in action?

Interactivity has quite a few markdown code blocks. As it stands, the two different code blocks have different styles though :/ 

> should we make this something that we can opt out? It would require creating a new line in head partial and a new data point highlightJs in context that can be turned off. Putting next to the css feels a bit of a hack ;)

Yeah yeah yeah, we can add that when cleaning up at the end ;)

> (Btw not even sure if this is ready for review...)

It's absolutely ready for review :) And changes, and a new review after them ;)

**Comment by HugoGranstrom** [comment]
Hold on `hljs` is the only thing we have in the CSS file, so `nim` is the one we don't need. Trying to add `hljs` back, seems like it doesn't break anymore locally :rofl:

**Comment by pietroppeter** [comment]
> Interactivity has quite a few markdown code blocks. As it stands, the two different code blocks have different styles though 

I have looked now and it seems it comes out fine.

I would keep both nim and hljs in the class since nim is a semantic element (and hljs needed for css). And nohighlite should be enough to making it avoid doing the work twice (have not looked if there are error in console though).

I see the tests are failing but that is because we hard coded the nbCode expected output and we changed it.

**Comment by pietroppeter** [comment]
nbShow looks fine!

**Comment by HugoGranstrom** [comment]
The only difference I have found so far is that we are highlighting common functions in the static highlighting that highlight.js is not (above is static, below is dynamic highlighting):
image
I don't think this difference matters that much tbh.

> I would keep both nim and hljs in the class since nim is a semantic element (and hljs needed for css). And nohighlite should be enough to making it avoid doing the work twice (have not looked if there are error in console though).

Ok, will keep them then :+1:

File: docsrc/penguins.nim [hunk 1]
```diff
@@ -17,7 +17,10 @@ nbCode:
   let df = readCsv("data/penguins.csv")
 nbText: "let us see how it looks"
 nbCode:
-  echo df # why so wide the first column?
+  echo df
+nbText: "or even nicer using `nbShow(df)`:"
```
Comment (pietroppeter) [inline docsrc/penguins.nim:21]:
```suggestion
nbText: "or even nicer using `nbShow(df.head(10))`:"
```
for consistency with what we actually run

**Review by pietroppeter — APPROVED** [review]
Apart from a minor suggestion, it looks good!

One last detail:
- the highlight.min.js where is it taken from? from the past? If so I think in the past I selected a number of languages that are supported, not only nim (I see python and rust in the code, for example) I think I might have put the list in the documentation in the past.
- we probably should mention somewhere this new highlight feature (available not only for nim but for other languages) somewhere in the docs

**Comment by HugoGranstrom** [comment]
> the highlight.min.js where is it taken from? from the past? If so I think in the past I selected a number of languages that are supported, not only nim (I see python and rust in the code, for example) I think I might have put the list in the documentation in the past.

I took `common + Nim + Julia + Latex`, didn't really see any other language of interest on the list. If we want to add a new one in the future, it's just a matter of generating and uploading a new js file. 

> we probably should mention somewhere this new highlight feature (available not only for nim but for other languages) somewhere in the docs

Good idea :+1:  The question is where to put it, the readme probably?

**Comment by pietroppeter** [comment]
> > the highlight.min.js where is it taken from? from the past? If so I think in the past I selected a number of languages that are supported, not only nim (I see python and rust in the code, for example) I think I might have put the list in the documentation in the past.
> 
> I took `common + Nim + Julia + Latex`, didn't really see any other language of interest on the list. If we want to add a new one in the future, it's just a matter of generating and uploading a new js file.

ah good selection.

> 
> > we probably should mention somewhere this new highlight feature (available not only for nim but for other languages) somewhere in the docs
> 
> Good idea 👍 The question is where to put it, the readme probably?

yes, I would go with readme, a new subsection inside "Features" about "highlighting". There is already a mention of static highlighting in the bullet list at the beginning of features, that might be reviewed. Having the list of languages (even as a list to common in highlightjs) would be useful for the user.

**Comment by pietroppeter** [comment]
> Anything more you want to add?

nope, that's good, thanks for working on this. merging!

Next step would be to release with some appropriate notes, after that we add a commit to bump version and update changelog.

**Comment by HugoGranstrom** [comment]
> Next step would be to release with some appropriate notes, after that we add a commit to bump version and update changelog.

I'm on it!

**Comment by HugoGranstrom** [comment]
Have I forgotten something? For some reason it didn't allow me to create a discussion, something about there not existsing a title and description :shrug:

**Comment by HugoGranstrom** [comment]
> It usually tells you that you need to create a tag to populate automatic release notes

I'm not sure I follow, there were no problems generating the automatic release notes or creating the tag. It was just creating the discussion forum post that was the problem.

**Comment by pietroppeter** [comment]
Ah ok, not sure why then. Maybe you need special permissions?

**Comment by HugoGranstrom** [comment]
Yeah perhaps, not sure exactly what permissions I have. That should hopefully be easier to manage in nimib-org hopefully.

**Pull request merged**